libraries/AP_Baro/AP_Baro_BMP280.cpp

```diff
@@ -0,0 +1,172 @@
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "AP_Baro_BMP280.h"
+
+#include <utility>
+
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL &hal;
+
+#define BMP280_MODE_SLEEP  0
+#define BMP280_MODE_FORCED 1
+#define BMP280_MODE_NORMAL 3
+#define BMP280_MODE BMP280_MODE_NORMAL
+
+#define BMP280_OVERSAMPLING_1  1
+#define BMP280_OVERSAMPLING_2  2
+#define BMP280_OVERSAMPLING_4  3
+#define BMP280_OVERSAMPLING_8  4
+#define BMP280_OVERSAMPLING_16 5
+#define BMP280_OVERSAMPLING_P BMP280_OVERSAMPLING_16
+#define BMP280_OVERSAMPLING_T BMP280_OVERSAMPLING_2
+
+#define BMP280_FILTER_COEFFICIENT 2
+
+#define BMP280_ID            0x58
+
+#define BMP280_REG_ID        0xD0
+#define BMP280_REG_RESET     0xE0
+#define BMP280_REG_STATUS    0xF3
+#define BMP280_REG_CTRL_MEAS 0xF4
+#define BMP280_REG_CONFIG    0xF5
+#define BMP280_RED_DATA      0xF7
+
+AP_Baro_BMP280::AP_Baro_BMP280(AP_Baro &baro, AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+    : AP_Baro_Backend(baro)
+    , _dev(std::move(dev))
+{
+    uint8_t buf[24];
+
+    // Get pointer to I2C bus semaphore
+    AP_HAL::Semaphore *sem = _dev->get_semaphore();
+
+    // Take I2C bus sempahore
+    if (!sem->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        AP_HAL::panic("BMP280: unable to get semaphore");
+    }
+
+    // Check BMP280 ID
+    uint8_t val = 0;
+    
+    if (!_dev->read_registers(BMP280_REG_ID, &val, 1)) {
+        AP_HAL::panic("BMP280: can't read ID");
+    }
+    if (val != BMP280_ID) {
+        AP_HAL::panic("BMP280: wrong ID");
+    }
+
+    // Read the calibration data registers
+    if (!_dev->read_registers(0x88, buf, sizeof(buf))) {
+        AP_HAL::panic("BMP280: bad calibration registers");
+    }
+
+    _t1 = ((int16_t)buf[1] << 8) | buf[0];
+    _t2 = ((int16_t)buf[3] << 8) | buf[2];
+    _t3 = ((int16_t)buf[5] << 8) | buf[4];
+    _p1 = ((int16_t)buf[7] << 8) | buf[6];
+    _p2 = ((int16_t)buf[9] << 8) | buf[8];
+    _p3 = ((int16_t)buf[11] << 8) | buf[10];
+    _p4 = ((int16_t)buf[13] << 8) | buf[12];
+    _p5 = ((int16_t)buf[15] << 8) | buf[14];
+    _p6 = ((int16_t)buf[17] << 8) | buf[16];
+    _p7 = ((int16_t)buf[19] << 8) | buf[18];
+    _p8 = ((int16_t)buf[21] << 8) | buf[20];
+    _p9 = ((int16_t)buf[23] << 8) | buf[22];
+
+    _dev->write_register(BMP280_REG_CTRL_MEAS, (BMP280_OVERSAMPLING_T << 5) + (BMP280_OVERSAMPLING_P << 2) + BMP280_MODE);
+    _dev->write_register(BMP280_REG_CONFIG, BMP280_FILTER_COEFFICIENT << 2);
+
+    _instance = _frontend.register_sensor();
+
+    sem->give();
+
+    // Request 50Hz update
+    _dev->register_periodic_callback(20 * USEC_PER_MSEC, FUNCTOR_BIND_MEMBER(&AP_Baro_BMP280::_timer, bool));
+}
+
+//  Acumulate a new sensor reading
+bool AP_Baro_BMP280::_timer(void)
+{
+    uint8_t buf[6];
+    
+    if (!_dev->read_registers(BMP280_RED_DATA, buf, sizeof(buf))) {
+        _dev->set_speed(AP_HAL::Device::SPEED_LOW);
+        return false;
```

Comment:
1) don't set the bus speed in case of errors.... this is intended for devices that operate on different speeds depending on the register we are accessing
2) If you return false here you are basically removing the callback from the "periodic callback infra", i.e. you won't be called anymore even though the baro is still registered. Failing a few transfers is not desirable , but it's better to keep trying  (it may occur because of interference on the I2C lines).